Service worker support

maier49 opened the pull request:
**Type:** bug / feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

<!--
Our bots should ensure:

* [ ] All contributors have signed a CLA
* [ ] The PR passes CI testing
* [ ] Code coverage is maintained
* [ ] The PR has been reviewed and approved
-->

**Description:**

Relates to dojo/meta#169. Adds a flag `--serviceWorker`. If this flag is specified, and not running in watch mode or building tests, the `process.env.DOJO_SERVICE_WORKER` flag will be set to true, and the `sw-precache-webpack-plugin` will be enabled, adding support for caching all resources with a service worker to the app.  

The [hacker news PWA PR](https://github.com/dojo/examples/pull/227), which is still a WIP, if built with this branch will work offline. With this and the changes it depends on in `dojo/stores` for IndexedDB the entire app currently runs offline after the first load.

matt-gadd commented on package.json:
```diff
@@ -46,6 +46,7 @@
     "mockery": "^1.7.0",
     "remap-istanbul": "^0.6.4",
     "sinon": "^1.17.5",
+    "sw-precache-webpack-plugin": "^0.11.3",
```
Shouldn't this be in `dependencies `?

maier49 replied:
Yes, fixed

matt-gadd commented on src/main.ts:
```diff
@@ -205,6 +206,11 @@ const command: Command = {
 			type: 'boolean'
 		});
+
+		options('serviceWorker', {
```
Can we move this to a `dojorc` config option instead.

maier49 replied:
What does that consist of specifically? The tests for dojorc that I've seen all involve loading arguments defined here from a configuration file.

maier49 replied:
I removed the call to `options()` so this will now only be supported via `dojorc`. As a follow up to that, it doesn't seem like we have anywhere to document non command line argument configuration options. Would adding a section to the README work?

matt-gadd commented on src/webpack.config.ts:
```diff
@@ -116,6 +122,9 @@ function webpackConfig(args: Partial<BuildArgs>) {
 				}
 			}),
+			new DefinePlugin({
+				'process.env.DOJO_SERVICE_WORKERS': JSON.stringify(includeServiceWorker)
```
Do we have a standard for env options? Should we put them all under one dojo namespace?

maier49 replied:
I didn't see anywhere else where we were using environment variables. Putting them all under a dojo namespace makes sense to me. `process.env.DOJO.SERVICE_WORKER`?

kitsonk replied:
We would need to move `DOJO.CLI` as well...  But agreed.

maier49 commented:
I addressed the comments here, with the exceptions that I didn't change DOJO_CLI because it's not set here, and I haven't tackled the more general testing issue.

maier49 commented:
@kitsonk @matt-gadd All the feedback here has been addressed and it's been rebased against the latest.

agubler commented:
@maier49 apologies seems we have conflicts again, can we resolve them so we can get this merged. Thanks!

maier49 commented:
In light of the concerns mentioned in [this comment](https://github.com/dojo/meta/issues/83#issuecomment-321250186), we will take a different approach for supporting service workers and other PWA features. It will involve using `.dojorc` and generating appropriate files in `cli-build-webpack`, as opposed to adding code to `cli-create-app`  that may not apply in all cases.
